parquet-variant/src/main/java/org/apache/parquet/variant/Variant.java

```diff
@@ -0,0 +1,444 @@
+  // `[0, objectSize())`.
+  // It is only legal to call it when `getType()` is `Type.OBJECT`.
+  /**
+   * Returns the object field at the `index` slot. Return null if `index` is out of the bound of
+   * `[0, objectSize())`. `getType()` must be `Type.OBJECT`.
+   * @param index the index of the object field to get
+   * @return the Objectfield at the `index` slot, or null if `index` is out of bounds
+   */
+  public ObjectField getFieldAtIndex(int index) {
+    return handleObject(value, pos, (size, idSize, offsetSize, idStart, offsetStart, dataStart) -> {
+      if (index < 0 || index >= size) return null;
+      int id = readUnsigned(value, idStart + idSize * index, idSize);
+      int offset = readUnsigned(value, offsetStart + offsetSize * index, offsetSize);
+      String key = getMetadataKey(metadata, id);
+      Variant v = new Variant(value, metadata, dataStart + offset);
+      return new ObjectField(key, v);
+    });
+  }
+
+  /**
+   * Returns the dictionary ID for the object field at the `index` slot.
+   * `getType()` must be `Type.OBJECT`.
+   * @param index the index of the object field to get the dictionary ID for
+   * @return the dictionary ID for the object field at the `index` slot
+   * @throws MalformedVariantException if `index` is out of bounds
+   */
+  public int getDictionaryIdAtIndex(int index) {
+    return handleObject(value, pos, (size, idSize, offsetSize, idStart, offsetStart, dataStart) -> {
+      if (index < 0 || index >= size) {
+        throw malformedVariant();
+      }
+      return readUnsigned(value, idStart + idSize * index, idSize);
+    });
+  }
+
+  /**
+   * @return the number of array elements. `getType()` must be `Type.ARRAY`.
+   */
+  public int arraySize() {
+    return handleArray(value, pos, (size, offsetSize, offsetStart, dataStart) -> size);
+  }
+
+  /**
+   * Returns the array element Variant value at the `index` slot. Returns null if `index` is
+   * out of the bound of `[0, arraySize())`. `getType()` must be `Type.ARRAY`.
+   * @param index the index of the array element to get
+   * @return the array element Variant at the `index` slot, or null if `index` is out of bounds
+   */
+  public Variant getElementAtIndex(int index) {
+    return handleArray(value, pos, (size, offsetSize, offsetStart, dataStart) -> {
+      if (index < 0 || index >= size) return null;
+      int offset = readUnsigned(value, offsetStart + offsetSize * index, offsetSize);
+      return new Variant(value, metadata, dataStart + offset);
+    });
+  }
+
+  /**
+   * @param zoneId The ZoneId to use for formatting timestamps
+   * @return the JSON representation of the variant
+   * @throws MalformedVariantException if the variant is malformed
+   */
+  public String toJson(ZoneId zoneId) {
+    return toJson(zoneId, false);
+  }
+
+  /**
+   * @param zoneId The ZoneId to use for formatting timestamps
+   * @param truncateTrailingZeros Whether to truncate trailing zeros in decimal values or timestamps
```

Comment:
I don't think this is allowed by the JSON conversion spec either.

Reply:
We've removed the JSON conversion requirements from the spec, so this is no longer relevant.